pkg/storage/cacher.go

```diff
@@ -0,0 +1,327 @@
+	// Maximum size of the history cached in memory.
+	CacheCapacity int
+
+	// An underlying storage.Interface.
+	Storage Interface
+
+	// The Cache will be caching objects of a given Type and assumes that they
+	// are all stored under ResourcePrefix directory in the underlying database.
+	Type           interface{}
+	ResourcePrefix string
+
+	// KeyFunc is used to get a key in the underyling storage for a given object.
+	KeyFunc func(runtime.Object) (string, error)
+
+	// NewList is a function that creates new empty object storing a list of
+	// objects of type Type.
+	NewListFunc func() runtime.Object
+}
+
+// Cacher is responsible for serving WATCH and LIST requests for a given
+// resource from its internal cache and updating its cache in the background
+// based on the underlying storage contents.
+type Cacher struct {
+	sync.RWMutex
+
+	// Whether Cacher is initialized.
+	initialized sync.WaitGroup
+	initOnce    sync.Once
+
+	// "sliding window" of recent changes of objects and the current state.
+	watchCache *cache.WatchCache
+	reflector  *cache.Reflector
+
+	// Registered watchers.
+	watcherIdx int
+	watchers   map[int]*cacheWatcher
+
+	// Versioner is used to handle resource versions.
+	versioner Versioner
+
+	// keyFunc is used to get a key in the underyling storage for a given object.
+	keyFunc func(runtime.Object) (string, error)
+}
+
+func NewCacher(config CacherConfig) *Cacher {
+	watchCache := cache.NewWatchCache(config.CacheCapacity)
+	listerWatcher := newCacherListerWatcher(config.Storage, config.ResourcePrefix, config.NewListFunc)
+
+	cacher := &Cacher{
+		initialized: sync.WaitGroup{},
+		watchCache:  watchCache,
+		reflector:   cache.NewReflector(listerWatcher, config.Type, watchCache, 0),
+		watcherIdx:  0,
+		watchers:    make(map[int]*cacheWatcher),
+		versioner:   APIObjectVersioner{},
+		keyFunc:     config.KeyFunc,
+	}
+	cacher.initialized.Add(1)
+	// See startCaching method for why explanation on it.
+	watchCache.SetOnReplace(func() { cacher.initOnce.Do(func() { cacher.initialized.Done() }); cacher.Unlock() })
+	watchCache.SetOnEvent(cacher.processEvent)
+
+	stopCh := util.NeverStop
+	go util.Until(func() { cacher.startCaching(stopCh) }, 0, stopCh)
+	cacher.initialized.Wait()
+	return cacher
+}
+
+func (c *Cacher) startCaching(stopChannel <-chan struct{}) {
+	c.Lock()
+	c.terminateAllWatchers()
+	// We explicitly do NOT Unlock() in this method.
+	// This is because we do not want to allow any WATCH/LIST methods before
+	// the cache is initialized. Once the underlying cache is propagated,
+	// onReplace handler will be called, which will do the Unlock() as
+	// configured in NewCacher().
+	// Note that since onReplace may be not called due to errors, we explicitly
+	// need to retry it on errors under lock.
+	for {
+		err := c.reflector.ListAndWatch(stopChannel)
+		if err == nil {
+			break
+		}
+	}
+}
+
+func (c *Cacher) Watch(key string, resourceVersion uint64, filter FilterFunc) (watch.Interface, error) {
```

Comment:
OK - I see your point.
On the other hand during my experiments with the new approach I didn't see any "out of the window" failure, so hopefully this would be very rare.